src/emqtt_inflight.erl

```diff
@@ -0,0 +1,151 @@
+       expire_at()
+      }).
+
+-type(seq_no() :: pos_integer()).
+-type(sent_at() :: non_neg_integer()). %% in millisecond
+-type(expire_at() :: non_neg_integer() | infinity). %% in millisecond
+
+-type(retry_func() :: fun(
+                        (emqtt:packet_id(),
+                         inflight_publish() | inflight_pubrel()
+                        ) -> inflight_publish() | inflight_pubrel())).
+
+-export_type([inflight/0, inflight_publish/0, inflight_pubrel/0,
+              sent_at/0, expire_at/0, retry_func/0]).
+
+%%--------------------------------------------------------------------
+%% APIs
+
+-spec(new(infinity | pos_integer()) -> inflight()).
+new(MaxInflight) ->
+    #{max_inflight => MaxInflight, sent => #{}, size => 0, seq => 1}.
+
+insert(_Id, _Req, #{max_inflight := Max, size := Max}) ->
+    error;
+insert(Id, Req, Inflight = #{sent := Sent, size := Size, seq := Seq}) ->
+    Inflight#{sent := maps:put(Id, {Seq, Req}, Sent),
+              size := Size + 1,
+              seq  := Seq + 1
+             }.
+
+update(_Id, _Req, #{max_inflight := Max, size := Max}) ->
+    error;
+update(Id, Req, Inflight = #{sent := Sent}) ->
+    case maps:find(Id, Sent) of
+        error -> error;
+        {ok, {No, _OldReq}} ->
+            Inflight#{sent := maps:put(Id, {No, Req}, Sent)}
+    end.
+
+delete(_Id, _Inflight = #{size := 0}) ->
+    error;
+delete(Id, Inflight = #{sent := Sent, size := Size}) ->
+    case maps:take(Id, Sent) of
+        error -> error;
+        {{_, Req}, Sent1} ->
+            {Req, Inflight#{sent := Sent1, size := Size - 1}}
+    end.
+
+size(#{size := Size}) ->
+    Size.
+
+is_full(#{max_inflight := infinity}) ->
+    false;
+is_full(#{max_inflight := Max, size := Size}) ->
+    Max =:= Size.
+
+is_empty(#{size := Size}) ->
+    Size =< 0.
+
+%% @doc first in first evaluate
+foreach(_F, #{size := 0}) ->
+    ok;
+foreach(F, #{sent := Sent}) ->
+    lists:foreach(
+      fun({Id, {_SeqNo, Req}}) -> F(Id, Req) end,
+      arrange_sent(Sent)
+     ).
+
+%% @doc eval RetryFunc if held time(ms) over HeldIntv
+-spec(retry(pos_integer(), retry_func(), inflight())
+      -> list(inflight_publish() | inflight_pubrel())).
+retry(_HeldIntv, _RetryFun, Inflight = #{size := 0}) ->
+    Inflight;
+retry(HeldIntv, RetryFun, Inflight = #{sent := Sent}) ->
+    Now = erlang:system_time(millisecond),
+    Need = arrange_sent(
+             filter_sent(
+               fun(_Id, Req) -> (sent_at(Req) + HeldIntv) < Now end,
+               Sent
+              )),
+    NSent = lists:foldl(
+              fun({Id, {SeqNo, Req}}, SentAcc) ->
+                      NReq = RetryFun(Id, Req),
+                      maps:put(Id, {SeqNo, NReq}, SentAcc)
+              end, Sent, Need),
+    Inflight#{sent := NSent}.
+
+%%--------------------------------------------------------------------
+%% Internal funcs
+
+filter_sent(F, Sent) ->
+    maps:filter(fun(Id, {_SeqNo, Req}) -> F(Id, Req) end, Sent).
+
+%% XXX: break opaque?
```

Comment:
if we move retry/4 to emqtt.erl, the inflight items can be an opaque object?
it seems we only need to export the `arrange_sent` (suggested to rename to `sort_sent`) API?

return `[{ID, Req}]` instead of `[{ID, {Seqno, Req}}]` from the `sort_sent` API.
then the caller can make perform retry folding through this list,
and call `update` API to update the sent timestamp.